src/appleseed/renderer/kernel/lighting/pathtracer.h

```diff
@@ -178,7 +178,8 @@ size_t PathTracer<PathVisitor, Adjoint>::trace(
 
         // Retrieve the ray.
         const ShadingRay& ray = vertex.get_ray();
-        vertex.m_outgoing = foundation::normalize(-ray.m_dir);
+        assert(foundation::is_normalized(ray.m_dir));
```

Comment:
I would suggest checking the length of the ray direction in the constructors of `ShadingRay`. Then this assertion (and probably a few others) could be removed.